key.go

```diff
@@ -0,0 +1,810 @@
+	KeyTypeEc2 KeyType = 2
+	// Symmetric Keys
+	KeyTypeSymmetric KeyType = 4
+)
+
+// KeyTypeFromString returns the KeyType corresponding to the specified name.
+func KeyTypeFromString(v string) (KeyType, error) {
+	switch v {
+	case "OKP":
+		return KeyTypeOkp, nil
+	case "EC2":
+		return KeyTypeEc2, nil
+	case "Symmetric":
+		return KeyTypeSymmetric, nil
+	default:
+		return KeyTypeInvalid, fmt.Errorf("unknown key type value %q", v)
+	}
+}
+
+func (kt KeyType) String() string {
+	switch kt {
+	case KeyTypeOkp:
+		return "OKP"
+	case KeyTypeEc2:
+		return "EC2"
+	case KeyTypeSymmetric:
+		return "Symmetric"
+	default:
+		return "unknown key type value " + strconv.Itoa(int(kt))
+	}
+}
+
+// MarshalCBOR marshals the KeyType as a CBOR int.
+func (kt KeyType) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(kt))
+}
+
+// UnmarshalCBOR populates the KeyType from the provided CBOR value (must be
+// int or tstr).
+func (kt *KeyType) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key type value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := KeyTypeFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*kt = v
+	} else {
+		v := raw.Int()
+
+		if v == 0 {
+			// 0  is reserved, and so can never be valid
+			return fmt.Errorf("invalid key type value 0")
+		}
+
+		if v > 4 || v < 0 || v == 3 {
+			return fmt.Errorf("unknown key type value %d", v)
+		}
+
+		*kt = KeyType(v)
+	}
+
+	return nil
+}
+
+const (
+
+	// Invalid/unrecognised curve
+	CurveInvalid Curve = 0
+
+	// NIST P-256 also known as secp256r1
+	CurveP256 Curve = 1
+
+	// NIST P-384 also known as secp384r1
+	CurveP384 Curve = 2
+
+	// NIST P-521 also known as secp521r1
+	CurveP521 Curve = 3
+
+	// X25519 for use w/ ECDH only
+	CurveX25519 Curve = 4
+
+	// X448 for use w/ ECDH only
+	CurveX448 Curve = 5
+
+	// Ed25519 for use /w EdDSA only
+	CurveEd25519 Curve = 6
+
+	// Ed448 for use /w EdDSA only
+	CurveEd448 Curve = 7
+)
+
+// Curve represents the EC2/OKP key's curve. See:
+// https://datatracker.ietf.org/doc/html/rfc8152#section-13.1
+type Curve int64
+
+func CurveFromString(v string) (Curve, error) {
+	switch v {
+	case "P-256":
+		return CurveP256, nil
+	case "P-384":
+		return CurveP384, nil
+	case "P-521":
+		return CurveP521, nil
+	case "X25519":
+		return CurveX25519, nil
+	case "X448":
+		return CurveX448, nil
+	case "Ed25519":
+		return CurveEd25519, nil
+	case "Ed448":
+		return CurveEd448, nil
+	default:
+		return CurveInvalid, fmt.Errorf("unknown curve value %q", v)
+	}
+}
+
+func (c Curve) String() string {
+	switch c {
+	case CurveP256:
+		return "P-256"
+	case CurveP384:
+		return "P-384"
+	case CurveP521:
+		return "P-521"
+	case CurveX25519:
+		return "X25519"
+	case CurveX448:
+		return "X448"
+	case CurveEd25519:
+		return "Ed25519"
+	case CurveEd448:
+		return "Ed448"
+	default:
+		return "unknown curve value " + strconv.Itoa(int(c))
+	}
+}
```

Comment:
Same for the `Curve`. The name field should not be considered as value. See [RFC 8152 Section 16.8](https://datatracker.ietf.org/doc/html/rfc8152#section-16.8):

>    Name:  This is a descriptive name that enables easier reference to
>       the item.  **It is not used in the encoding.**
> 
>    Value:  This is the value used to identify the curve.  These values
>       MUST be unique.  The integer values from -256 to 255 are
>       designated as "Standards Action".  The integer values from 256 to
>       65535 and -65536 to -257 are designated as "Specification
>       Required".  Integer values over 65535 are designated as "Expert
>       Review".  Integer values less than -65536 are marked as private
>       use.